File: app/src/main/java/com/dumuzeyn/mp3player/QueueAdapter.java
package com.dumuzeyn.mp3player;

import android.view.ViewGroup;
import android.widget.FrameLayout;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;
import java.util.List;

final class QueueAdapter extends RecyclerView.Adapter<QueueAdapter.Holder> {
    interface Listener {
        void remove(int index);
        void play(int index);
        void move(int from, int to);
    }

    private final MainActivityCore host;
    private final Listener listener;
    private final ArrayList<Track> tracks;

    QueueAdapter(MainActivityCore host, List<Track> tracks, Listener listener) {
        this.host = host;
        this.listener = listener;
        this.tracks = new ArrayList<>(tracks);
        setHasStableIds(true);
    }

    @Override
    public long getItemId(int position) {
        return tracks.get(position).trackId.hashCode();
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        FrameLayout container = new FrameLayout(host);
        container.setLayoutParams(new RecyclerView.LayoutParams(-1, -2));
        return new Holder(container);
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position) {
        holder.container.removeAllViews();
        Track track = tracks.get(position);
        holder.container.addView(host.songsRenderer.queueRow(track,
                () -> remove(holder.getBindingAdapterPosition()),
                () -> listener.play(holder.getBindingAdapterPosition())),
                new FrameLayout.LayoutParams(-1, -2));
    }

    @Override
    public int getItemCount() {
        return tracks.size();
    }

    boolean move(int from, int to) {
        if (from < 0 || to < 0 || from >= tracks.size() || to >= tracks.size()) {
            return false;
        }
        Track moved = tracks.remove(from);
        tracks.add(to, moved);
        notifyItemMoved(from, to);
        listener.move(from, to);
        return true;
    }

    void remove(int position) {
        if (position < 0 || position >= tracks.size()) {
            return;
        }
        tracks.remove(position);
        notifyItemRemoved(position);
        listener.remove(position);
    }

    static final class Holder extends RecyclerView.ViewHolder {
        final FrameLayout container;

        Holder(FrameLayout container) {
            super(container);
            this.container = container;
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/QueueAdapter.kt
package com.dumuzeyn.mp3player

import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView

internal class QueueAdapter(
    private val host: MainActivityCore,
    tracks: List<Track>,
    private val listener: Listener,
) : ListAdapter<Track, QueueAdapter.Holder>(diff) {
    interface Listener {
        fun remove(index: Int)
        fun play(index: Int)
        fun move(from: Int, to: Int)
    }

    private var currentTrackId = ""

    init {
        setHasStableIds(true)
        captureCurrentTrack()
        submitTracks(tracks)
    }

    override fun getItemId(position: Int): Long = getItem(position).trackId.hashCode().toLong()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        val container = FrameLayout(host).apply {
            layoutParams = RecyclerView.LayoutParams(-1, -2)
        }
        return Holder(container)
    }

    override fun onBindViewHolder(holder: Holder, position: Int) {
        holder.container.removeAllViews()
        val track = getItem(position)
        holder.container.addView(
            host.songsRenderer.queueRow(
                track,
                { remove(holder.bindingAdapterPosition) },
                {
                    listener.play(holder.bindingAdapterPosition)
                    host.uiHandler.postDelayed(::refreshPlayback, 80L)
                },
            ),
            FrameLayout.LayoutParams(-1, -2),
        )
    }

    fun submitTracks(tracks: List<Track>) {
        submitList(ArrayList(tracks), ::refreshPlayback)
    }

    fun refreshPlayback() {
        val previous = currentTrackId
        captureCurrentTrack()
        notifyTrack(previous)
        if (previous != currentTrackId) notifyTrack(currentTrackId)
    }

    fun move(from: Int, to: Int): Boolean {
        if (from < 0 || to < 0 || from >= itemCount || to >= itemCount) return false
        val tracks = ArrayList(currentList)
        val moved = tracks.removeAt(from)
        tracks.add(to, moved)
        submitList(tracks)
        listener.move(from, to)
        return true
    }

    fun remove(position: Int) {
        if (position < 0 || position >= itemCount) return
        val tracks = ArrayList(currentList)
        tracks.removeAt(position)
        submitList(tracks)
        listener.remove(position)
    }

    private fun captureCurrentTrack() {
        currentTrackId = host.playbackStateProvider.currentTrack()?.trackId.orEmpty()
    }

    private fun notifyTrack(trackId: String) {
        if (trackId.isEmpty()) return
        currentList.forEachIndexed { index, track ->
            if (trackId == track.trackId) {
                notifyItemChanged(index)
                return
            }
        }
    }

    class Holder(val container: FrameLayout) : RecyclerView.ViewHolder(container)

    private companion object {
        val diff = object : DiffUtil.ItemCallback<Track>() {
            override fun areItemsTheSame(oldItem: Track, newItem: Track): Boolean =
                oldItem.trackId == newItem.trackId

            override fun areContentsTheSame(oldItem: Track, newItem: Track): Boolean =
                oldItem.uri == newItem.uri && oldItem.title == newItem.title
        }
    }
}
